mmv1/third_party/terraform/services/container/resource_container_node_pool.go.tmpl

```diff
@@ -1688,13 +1672,24 @@ func containerNodePoolAwaitRestingState(config *transport_tpg.Config, name, proj
 	return state, err
 }
 
-// Retries an operation while the canonical error code is FAILED_PRECONDTION
-// or RESOURCE_EXHAUSTED which indicates there is an incompatible operation
-// already running on the cluster or there are the number of allowed
-// concurrent operations running on the cluster. These errors can be safely
-// retried until the incompatible operation completes, and the newly
-// requested operation can begin.
-func retryWhileIncompatibleOperation(timeout time.Duration, lockKey string, f func() error) error {
+// Retries an operation while the canonical error code is FAILED_PRECONDTION or RESOURCE_EXHAUSTED which indicates
+// there is an incompatible operation already running on the cluster or there are the number of allowed concurrent
+// operations running on the cluster. These errors can be safely retried until the incompatible operation completes,
+// and the newly requested operation can begin.
+// The cluster lock is held during createOpFunc to make opeation creations sequencial, and cluster read lock is held
+// during waitOpFunc to allow concurrency.
+func retryWhileIncompatibleOperation(timeout time.Duration, lockKey string, clusterLockKey string, createOpFunc func() (*container.Operation, error), waitOpFunc func(*container.Operation) error) error {
```

Comment:
```suggestion
// Retries an operation while the canonical error code is FAILED_PRECONDTION or RESOURCE_EXHAUSTED which indicates
// there is an incompatible operation already running on the cluster or there are the number of allowed concurrent
// operations running on the cluster. These errors can be safely retried until the incompatible operation completes,
// and the newly requested operation can begin.
// The npLockKey is held during createOpFunc to make operations sequential within the node pool, and
// clusterLockKey is held during waitOpFunc to allow concurrency on a cluster.
func retryWhileIncompatibleOperation(timeout time.Duration, npLockKey string, clusterLockKey string, createOpFunc func() (*container.Operation, error), waitOpFunc func(*container.Operation) error) error {
```

Reply:
I probably should not have committed the suggestion because it was incorrect.
npLock is held through out this function, just for mutex reason. cluster WRITE lock is held during createOpFunc for sequential creation, and cluster READ lock is held during waitOpFunc to allow concurrency.

Reply:
ah ! sorry about that!